dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/credentials/CredentialsExpiryAlertTask.java

```diff
@@ -0,0 +1,161 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.hisp.dhis.message.MessageSender;
+import org.hisp.dhis.setting.SettingKey;
+import org.hisp.dhis.setting.SystemSettingManager;
+import org.hisp.dhis.system.util.DateUtils;
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.user.UserCredentials;
+import org.hisp.dhis.user.UserService;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import javax.annotation.Resource;
+import java.util.*;
+
+/**
+ * Created by zubair on 29.03.17.
+ */
+public class CredentialsExpiryAlertTask implements Runnable
+{
+    private static final Log log = LogFactory.getLog( CredentialsExpiryAlertTask.class );
+
+    private static final String SUBJECT = "Password Expiry Alert";
+
+    private static final String TEXT = "Dear %s, Please change your password. It will expire in %d days.";
+
+    public static final String KEY_TASK = "credentialsExpiryAlertTask";
+
+    private static int FIRST_EXPIRY_THRESHOLD = 14;
+
+    private static int SECOND_EXPIRY_THRESHOLD = 7;
+
+    // -------------------------------------------------------------------------
+    // Dependencies
+    // -------------------------------------------------------------------------
+
+    @Autowired
+    private UserService userService;
+
+    @Autowired
+    @Resource ( name = "emailMessageSender" )
+    private MessageSender emailMessageSender;
+
+    @Autowired
+    private SystemSettingManager systemSettingManager;
+
+    // -------------------------------------------------------------------------
+    // Implementation
+    // -------------------------------------------------------------------------
+
+    @Override
+    public void run()
+    {
+        boolean isExpiryAlertEnabled = (Boolean) systemSettingManager.getSystemSetting( SettingKey.CREDENTIALS_EXPIRY_ALERT );
+
+        if ( !isExpiryAlertEnabled )
+        {
+            log.info( String.format( "%s aborted. Expiry alerts are disabled", KEY_TASK ) );
+
+            return;
+        }
+
+        log.info( String.format( "%s has started", KEY_TASK ) );
+
+        List<User> users = userService.getAllUsers();
```

Comment:
Let's do this in HQL instead, you can do the date calculation in HQL and retrieve only users which is about to expire. Getting all users might caused out-of-memory as some installations have tens of thousands of users.